File: DesignPatterns/src/C++/001-SOLID/002-SRP/Modem/Modem.cpp
#include "Modem.hpp"

#include <iostream>

void Modem::connect()
{
	std::cout << "Modem::connect" << std::endl;
}

void Modem::disconnect()
{
	std::cout << "Modem::Disconnect" << std::endl;
}

size_t Modem::send(const char *buf, size_t size)
{
	std::cout << "Modem::send" << std::endl;

	return size;
}
size_t Modem::receive(char *buf, size_t size)
{
	std::cout << "Modem::receive" << std::endl;

	return size;
}

File: src/DesignPatterns/src/Java/001-SOLID/001-SRP/ModemApp/src/org/csystem/app/modemapp/Modem.java
package org.csystem.app.modemapp;

import java.io.Closeable;
import java.io.IOException;

public class Modem implements Closeable {
    private IConnection m_connection;
    private ITransmission m_transmission;
    //...

    public Modem(IConnection connection, ITransmission transmission)
    {
        m_connection = connection;
        m_transmission = transmission;
        m_connection.connect();
    }

    //...

    public void sendMessage(String msg)
    {
        m_transmission.send(msg);
    }

    public String receiveMessage()
    {
        return m_transmission.receive();
    }

    @Override
    public void close() throws IOException
    {
        m_connection.disconnect();
    }
}
